aiocouch/attachment.py

```diff
@@ -0,0 +1,57 @@
+# Copyright (c) 2020, Adrien Vergé
+#
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without modification,
+# are permitted provided that the following conditions are met:
+#
+#     * Redistributions of source code must retain the above copyright notice,
+#       this list of conditions and the following disclaimer.
+#     * Redistributions in binary form must reproduce the above copyright notice,
+#       this list of conditions and the following disclaimer in the documentation
+#       and/or other materials provided with the distribution.
+#     * Neither the name of metricq nor the names of its contributors
+#       may be used to endorse or promote products derived from this software
+#       without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+# NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from .remote import RemoteAttachment
+
+
+class Attachment(RemoteAttachment):
+    async def exists(self):
+        return await self._exists()
+
+    async def fetch(self):
+        return await self._get()
+
+    async def save(self, data, content_type):
+        data = await self._put(await self._get_doc_rev(), data, content_type)
+        # Parent document needs to have '_attachments' and '_rev' updated:
+        await self._document.fetch()
+        self.content_type = content_type
```

Comment:
Maybe also move these to the `RemoteAttachment` class

Reply:
Yes, I hesitated when doing the previous change... I have to say it's hard to see the need for 2 classes, now `Attachment` is pretty much a wrapper for `RemoteAttachment` with no real code in it. But I guess for other classes that distinction makes sense, so let's keep the same structure for attachments.

Anyway, now it's done, I just pushed again.

Reply:
Maybe I should share my thoughts on this split. While the `Attachment` class is indeed a very thin wrapper for the `RemoteAttachment` class, there is still a sharp division in concerns between both. `Attachment` is the interface that aiocouch presents, but `RemoteAttachment` is a wrapper of the CouchDB interface. Having said that I'm actually undecided, where the `content_type` member should reside. There are good arguments for both classes.

However, it should be consistent and it seems you left some in both classes.

Reply:
This public/private splitting makes sense, that's what I actually guessed. But for example `content_type` is a "public-side" property, that's why I initially put it in `Attachment`.

>  it seems you left some in both classes.

Sorry about that, fixed:
```diff
--- a/aiocouch/attachment.py
+++ b/aiocouch/attachment.py
@@ -38,13 +38,11 @@ class Attachment(RemoteAttachment):
 
     async def save(self, data, content_type):
         await self._put(await self._get_doc_rev(), data, content_type)
-        self.content_type = content_type
         # Parent document needs to have '_attachments' and '_rev' updated:
         await self._document.fetch()
 
     async def delete(self):
         await self._delete(await self._get_doc_rev())
-        self.content_type = None
         # Parent document needs to have '_attachments' and '_rev' updated:
         await self._document.fetch()
 
```

Reply:
> But for example content_type is a "public-side" property, that's why I initially put it in Attachment.

Yeah, that's the argument to put it there. But it's also related to networking and CouchDB interfacing, so either way is okay, I think.

Anyways, thanks for the work.